internal/npm_install/generate_build_file.ts

```diff
@@ -571,53 +597,51 @@ export function getDirectDependencySet(pkgJsonPath: string): Set<string> {
 /**
  * Finds and returns an array of all packages under a given path.
  */
-function findPackages(p: string, dependencies: Set<string>) {
-  if (!isDirectory(p)) {
-    return [];
+async function findPackages(p: string, dependencies: Set<string>, pkgs: Array<Dep>) {
+  if (!await isDirectory(p)) {
+    return;
   }
 
-  const pkgs: Dep[] = [];
-
-  const listing = fs.readdirSync(p);
-
-  const packages = listing
-                       // filter out scopes
-                       .filter(f => !f.startsWith('@'))
-                       // filter out folders such as `.bin` which can create
-                       // issues on Windows since these are "hidden" by default
-                       .filter(f => !f.startsWith('.'))
-                       .map(f => path.posix.join(p, f))
-                       .filter(f => isDirectory(f));
-
-  packages.forEach(f => {
-    pkgs.push(parsePackage(f, dependencies), ...findPackages(path.posix.join(f, 'node_modules'), dependencies));
-  });
-
-  const scopes = listing.filter(f => f.startsWith('@'))
-                     .map(f => path.posix.join(p, f))
-                     .filter(f => isDirectory(f));
-  scopes.forEach(f => pkgs.push(...findPackages(f, dependencies)));
-
-  return pkgs;
+  const listing = await fs.readdir(p);
+
+  await Promise.all(listing.map(async f => {
+    // filter out folders such as `.bin` which can create
+    // issues on Windows since these are "hidden" by default
+    if (f.startsWith('.')) return;
+    const pf = path.posix.join(p, f)
+    
+    if (await isDirectory(pf)) {
+      if (f.startsWith('@')) {
+        await findPackages(pf, dependencies, pkgs)
+      } else {
+        pkgs.push(await parsePackage(pf, dependencies));
+        await findPackages(path.posix.join(pf, 'node_modules'), dependencies, pkgs)
+      }
+    }
+  }))
 }
 
 /**
  * Finds and returns an array of all package scopes in node_modules.
  */
-function findScopes() {
+async function findScopes() {
   const p = 'node_modules';
-  if (!isDirectory(p)) {
+  if (!await isDirectory(p)) {
     return [];
   }
 
-  const listing = fs.readdirSync(p);
-
-  const scopes = listing.filter(f => f.startsWith('@'))
-                     .map(f => path.posix.join(p, f))
-                     .filter(f => isDirectory(f))
-                     // strip 'node_modules/' from filename
-                     .map(f => f.substring('node_modules/'.length));
+  const listing = await fs.readdir(p);
 
+  const scopes: string[] = []
+  for (let f of listing) {
+    if (f.startsWith('@')) {
+      f = path.posix.join(p, f)
+      if (await isDirectory(f)) {
```

Comment:
Worth making this one concurrent? (putting the promises into `scopes` and then `Promise.all` instead of chaining them)

Reply:
I'm not sure, generate step didn't take much time.

Reply:
OK, use `Promise.all` now, you can check it.

Reply:
> I'm not sure, generate step didn't take much time.

If it makes the PR more complicated or you don't think it's worth it that's fine, we can always do it later.